### ISSUE TITLE
problem with vector objects

### ISSUE THREAD

**Issue by ThomasBreuer** [gap-system/gap#5030]
The following happens in the master branch and in GAP 4.12.0 but not in GAP 4.11.1.
```
gap> v:= [ Z(2)^0 ];;
gap> ConvertToVectorRep( v, GF(2) );;
gap> MTX.SpinnedBasis( v, [ [ [ Z(2)^0 ] ] ], GF(2^9) );
Error, no method found! For debugging hints type ?Recovery from NoMethodFound
Error, no 1st choice method found for `*' on 2 arguments
The 1st argument is 'fail' which might point to an earlier problem
[...]
```
This can be reduced to the following problem (in GAP 4.12.0 but not in GAP 4.11.1).
```
gap> F:= GF(2^9);;
gap> v:= [ Z(2) ];;
gap> ConvertToVectorRep( v, GF(2) );
2
gap> v:= ImmutableVector( F, v );
<immutable plist vector over GF(2^9) of length 1>
gap> m:= [ [ Z(2) ] ];;
gap> w:= v * m;
<immutable plist vector over GF(2^9) of length 1>
gap> w[1];
[ [ Z(2)^0 ] ]
```
The multiplication `v * m` is done via the method `for a plist vector and a scalar`, which does not check whether the arguments fit; we have to add checks in the code of this method or restrict its applicability.
In the code of `MTX.SpinnedBasis`, the problem seems to be that `ImmutableVector( F, v )` returns a vector object (a plist vector) but that the acting matrices do not get replaced by the corresponding matrix objects.
(In GAP 4.11.1, `ImmutableVector` returns a GF(2) vector not a plist vector.)

**Comment by ThomasBreuer** [comment]
I think the situation here is similar to that in #4983, where the question was raised whether `ImmutableMatrix` is allowed to convert between list of lists matrices and matrix objects. Here the question is whether `ImmutableVector is allowed to convert between vectors that are lists and vector objects. In both cases, the point is that the arithmetic behaviour of the two objects is different.

However, there is no complete analogy because the vector `v` that appears here as the result of `ConvertToVectorRep` is both a list and a vector object. (Is this perhaps the real problem?)

```
gap> F:= GF(2^9);;
gap> v:= [ Z(2) ];;
gap> ConvertToVectorRep( v, GF(2) );;
gap> IsList( v );
true
gap> IsVectorObj( v );
true
```

`IsVectorObj` is implied by `IsGF2VectorRep`, and `IsList` is set in the creation of `TYPE_LIST_GF2VEC`.
Thus one can argue that the `IsVectorObj` filter allows `ImmutableVector( F, v )` to return a "better" vector object `vv`, say, that is not in `IsList`.
Both `v` and `vv` can be multiplied with matrices, but they behave differently.

The methods implementing arithmetics for `IsPlistVectorRep` are a second problem, their installations are not restrictive enough.
For example,
```
InstallMethod( \*, "for a plist vector and a scalar",
  [ IsPlistVectorRep, IsScalar ],
  function( v, s )
    return Objectify( TypeObj(v),
             [v![BDPOS],PROD_LIST_SCL_DEFAULT(v![ELSPOS],s)] );
```
expects `PROD_LIST_SCL_DEFAULT` to create a list that is compatible with `v`, which does not follow from the requirement `[ IsPlistVectorRep, IsScalar ]`.

**Comment by frankluebeck** [comment]
I agree with this analysis. Just before the 4.12. release I removed various bugs, by making sure that `ImmutableVector` and `ImmutableMatrix` do not convert "classical" vectors/matrices into `IsVectorObj/IsMatrixObj` objects and vice versa. 
The problem in the example here is that objects in `IsGF2Vector/MatrixRep` or `Is8BitVector/MatrixRep` are both, "classical" and in `IsVector/MatrixObj`.  I'm not sure what the best solution is, maybe specific methods for these cases (which avoid to return `Vector/MatrixObj`s).

The method for multiplication shown in the previous post also caused some other problems which I have mentioned in the previous discussion. Mixed arithmetic between "classical" and `Vector/MatrixObj` objects must be avoided.

**Comment by ThomasBreuer** [comment]
@frankluebeck Thanks for the confirmation.
I think the (at least) two problems need independent fixes.

Concerning the first problem, my questions are as follows.
- Is it allowed at all that an object is in both `IsVectorObj` and `IsList`?
  `IsList` determines the arithmetic behaviour (here: multiplication with some matrix), and if `ImmutableVector` is allowed to choose a result that is not in `IsList` then we end up with a changed arithmetic behaviour.
- If the first question is answered with "yes" then is `ImmutableVector` allowed to return an object not in `IsList` if the argument is in `IsList`?
  If not then `ImmutableVector` will at least not make the situation worse.
  (Such a condition makes the documentation more complicated.)

Concerning the second problem, the methods for `IsPlistVectorRep` are intended to be fast, thus it is a bad idea to add checks to the code. It would be better to make the requirements of the installations more restrictive, but I do not know a good solution; `IsScalar` with no prescribed relation to the `IsPlistVectorRep` object is not enough.
